### ISSUE TITLE
Question about the Process Consistency Loss

### ISSUE THREAD

**Issue by shaform** [Lkydong2020/SEAL_WTAL#3]
Hi,

Thank you for the great work. I noticed that in the paper, the Process Consistency Loss is formulated as below,

<img width="505" height="99" alt="Image" src="https://github.com/user-attachments/assets/46125754-a71d-4fab-96ca-075a68454922" />.

I have some questions about the above equation. In particular,

1. It appears that the $$\delta(\dot)$$ function is not explicitly defined in the paper (?) Could you explain what this is?
2. The $$\varphi$$ function seems to be `mean squared error function` if we follow the definition in the earlier part of the paper. However, in your code, it appears that `mean squared error` is not being used. Could you explain the reason of the differences?
3. It appears that $$\frac{d_i}{I_i + \epsilon}$$ is the `normalized_center_diff` is the code? The code further normalize it using smooth_l1_loss, which seems to be different from what the paper describes. Could you explain the reason of the differences?

Thank you!

**Comment by Lkydong2020** [comment]
Hi,

Thank you very much for carefully reading both the paper and the code, and for raising these questions. Let me briefly clarify the relation between Eq. (10) and the implementation.

1. About $\delta(\cdot)$

   You are right that $\delta(\cdot)$ is not explicitly defined in the paper – this is our oversight.  
   Conceptually, $\delta(\mathcal{C}(i))$ is a penalty term that becomes non-zero when the overlap between a proposal and its pseudo-true instance is too small. In the code this is implemented as:

   p_gt_lengths = matched_p_gt_segments[:, 1] - matched_p_gt_segments[:, 0]
   intersection_ratio = intersection / p_gt_lengths
   penalty_indices = intersection_ratio < 0.5
   if penalty_indices.any():
       penalty_loss = (1 - pred_iou_selected[penalty_indices]).sum()
   else:
       penalty_loss = torch.tensor(0.0, device=device)

   So when the intersection ratio is less than 0.5, we add a penalty term (1 - predicted IoU). When the ratio is not less than 0.5, this term is set to zero. This corresponds to the $\delta(\mathcal{C}(i))$ term in Eq. (10).

2. About $\varphi$ (MSE vs Smooth L1)

   In the earlier part of the paper we described $\varphi$ as an MSE-type distance between predicted and pseudo-true centers.  
   During implementation we found that plain MSE was sensitive to outliers at early training stages, so we used Smooth L1 loss for better robustness (therefore, you can regard it as an engineering optimization that facilitates the training process.):

   pred_center_expr = (e1 - s2) / 2 * pred_iou_selected + (s1 + s2) / 2
   center_loss = F.smooth_l1_loss(
       pred_center_expr,
       matched_p_gt_centers.view(-1),
       reduction='sum'
   )

   This is the implementation of $\varphi(C_{P,i}, C_{G,i})$; the role is still a center-distance term, only the specific loss form is slightly changed for numerical stability. In fact, replacing it with MSE would not cause any problem and the final results would still demonstrate the effectiveness of our design.

3. About $d_i / (I_i + \epsilon)$ and normalized_center_diff

   Yes, normalized_center_diff in the code corresponds to the term $d_i / (I_i + \epsilon)$ in Eq. (10):

   center_diff = torch.abs(pred_centers[valid_indices] - matched_p_gt_centers)
   normalized_center_diff = center_diff / (intersection)  # intersection ≈ I_i
   normalized_center_diff_loss = F.smooth_l1_loss(
       normalized_center_diff,
       torch.zeros_like(normalized_center_diff),
       reduction='sum'
   )

   Here:
   - center_diff is $d_i$ (absolute center distance), and  
   - intersection is $I_i$ (intersection length, with a small clamp to avoid division by zero).

   In the paper we directly wrote the ratio $d_i / (I_i + \epsilon)$. In the code we additionally apply Smooth L1 to zero as a robust surrogate of the absolute value when aggregating this term.

Finally, the three components are combined as:

total_IOU_loss = 0.01 * total_center_loss \
                 + args.alpha_5 * total_normalized_center_loss \
                 + args.alpha_6 * total_penalty_loss

which corresponds to the three terms $\varphi(\cdot) + d_i / (I_i + \epsilon) + \delta(\cdot)$ in Eq. (10).

Thanks again for pointing this out !

Best regards,  
Li

**Comment by shaform** [comment]
Thanks a lot for your reply!
